js/src/jsapi.cpp

```diff
@@ -4927,114 +4927,138 @@ JS_ReportTaintSink(JSContext* cx, JS::HandleString str, const char* sink)
 {
   RootedValue arg(cx);
   arg.setNull();
-  JS_ReportTaintSink(cx, str, sink, arg);
+  JS::RootedValue handleVal(cx, JS::StringValue(str));
+  JS_ReportTaintSink(cx, handleVal, sink, arg);
 }
 
 JS_PUBLIC_API void
 JS_ReportTaintSink(JSContext* cx, JS::HandleValue value, const char* sink, JS::HandleValue arg)
 {
   if (value.isString()) {
     JSString *str = value.toString();
-    if (str) {
-      JS::RootedString strobj(cx, str);
-      JS_ReportTaintSink(cx, strobj, sink, arg);
+    if (!str)
+      return;
+
+    JS::RootedString strobj(cx, str);
+    if(!strobj->isTainted())
+      return;
+
+    // Extend the taint flow to include the sink function
+    strobj->taint().extend(TaintOperationFromContext(cx, sink, true, arg, true));
+
+    JS_ReportTaintSink(cx, sink, arg, strobj->taint().begin()->flow(), value, false);
+  }
+  else if(isTaintedNumber(value)){
+    NumberObject& number = value.toObject().as<NumberObject>();
+
+    // Extend the taint flow to include the sink function
+    number.taint().extend(TaintOperationFromContext(cx, sink, true, arg, true));
+
+    JS_ReportTaintSink(cx, sink, arg, number.taint(), value, true);
+
+  }
+  else if(value.isObject()){
```

Comment:
I see what you want to do here, but this could be fairly spammy. Is this missing a check for taintedness?